dspy/retrieve/pinecone_rm.py

```diff
@@ -33,6 +29,64 @@
 except Exception:
     ERRORS = (openai.RateLimitError, openai.APIError)
 
+
+class CloudEmbedProvider(ABC):
+    def __init__ (self, model, api_key=None):
+        self.model = model
+        self.api_key = api_key
+    
+    @abstractmethod
+    def get_embeddings(self, queries: List[str]) -> List[List[float]]:
+        pass
+
+class OpenAIEmbed(CloudEmbedProvider):
+    def __init__(self, model="text-embedding-ada-002", api_key: Optional[str]=None, org: Optional[str]=None):
+        super().__init__(model, api_key)
+        self.org = org
+        if self.api_key:
+            openai.api_key = self.api_key
+        if self.org:
+            openai.organization = org
+
+    
+    @backoff.on_exception(
+        backoff.expo,
+        ERRORS,
+        max_time=15,
+    )
+    def get_embeddings(self, queries: List[str]) -> List[List[float]]:
```

Comment:
could the caching functionality included [in this PR](https://github.com/stanfordnlp/dspy/pull/791/files#diff-c6ca43067290c41691d236de3c2a7648012ad0174adcb63f040e481a15f5b28bR171) be used here?

I realize we probably need an abstracted Retriever class to handle this but for now, feel free to check if this suffices